resolver/resolver.go

```diff
@@ -67,147 +69,206 @@ func (r *Resolver) ResolveToLastNode(ctx context.Context, fpath path.Path) (cid.
-		// able to resolve `zdpu[...]/foo`)
-		if lnk == nil {
-			break
-		}
-
-		if err != nil {
-			if err == dag.ErrLinkNotFound {
-				err = ErrNoLink{Name: p[0], Node: nd.Cid()}
-			}
-			return cid.Cid{}, nil, err
-		}
+	if len(nodes) < 1 {
+		return cid.Cid{}, nil, fmt.Errorf("path %v did not resolve to a node", fpath)
+	} else if len(nodes) < len(p) {
+		return cid.Undef, nil, ErrNoLink{Name: p[len(nodes)-1], Node: lastCid}
+	}
 
-		if len(rest) == 0 {
-			return lnk.Cid, nil, nil
-		}
+	parent := nodes[len(nodes)-1]
+	lastSegment := p[len(p)-1]
 
-		next, err := lnk.GetNode(ctx, r.DAG)
-		if err != nil {
-			return cid.Cid{}, nil, err
-		}
-		nd = next
-		p = rest
+	// find final path segment within node
+	nd, err := parent.LookupByString(lastSegment)
+	switch err.(type) {
+	case nil:
+	case schema.ErrNoSuchField:
+		return cid.Undef, nil, ErrNoLink{Name: lastSegment, Node: lastCid}
+	default:
+		return cid.Cid{}, nil, err
 	}
 
-	if len(p) == 0 {
-		return nd.Cid(), nil, nil
+	// if last node is not a link, just return it's cid, add path to remainder and return
+	if nd.Kind() != ipld.Kind_Link {
+		// return the cid and the remainder of the path
+		return lastCid, p[len(p)-depth-1:], nil
 	}
 
-	// Confirm the path exists within the object
-	val, rest, err := nd.Resolve(p)
+	lnk, err := nd.AsLink()
 	if err != nil {
-		if err == dag.ErrLinkNotFound {
-			err = ErrNoLink{Name: p[0], Node: nd.Cid()}
-		}
 		return cid.Cid{}, nil, err
 	}
 
-	if len(rest) > 0 {
-		return cid.Cid{}, nil, errors.New("path failed to resolve fully")
-	}
-	switch val.(type) {
-	case *ipld.Link:
-		return cid.Cid{}, nil, errors.New("inconsistent ResolveOnce / nd.Resolve")
-	default:
-		return nd.Cid(), p, nil
+	clnk, ok := lnk.(cidlink.Link)
+	if !ok {
+		return cid.Cid{}, nil, fmt.Errorf("path %v resolves to a link that is not a cid link: %v", fpath, lnk)
 	}
+
+	return clnk.Cid, []string{}, nil
 }
 
 // ResolvePath fetches the node for given path. It returns the last item
-// returned by ResolvePathComponents.
-func (r *Resolver) ResolvePath(ctx context.Context, fpath path.Path) (ipld.Node, error) {
+// returned by ResolvePathComponents and the last link traversed which can be used to recover the block.
+func (r *Resolver) ResolvePath(ctx context.Context, fpath path.Path) (ipld.Node, ipld.Link, error) {
 	// validate path
 	if err := fpath.IsValid(); err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
-	nodes, err := r.ResolvePathComponents(ctx, fpath)
-	if err != nil || nodes == nil {
-		return nil, err
+	c, p, err := path.SplitAbsPath(fpath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// create a selector to traverse all path segments but only match the last
+	pathSelector := pathLeafSelector(p)
+
+	nodes, c, _, err := r.resolveNodes(ctx, c, pathSelector)
+	if err != nil {
+		return nil, nil, err
+	}
+	if len(nodes) < 1 {
+		return nil, nil, fmt.Errorf("path %v did not resolve to a node", fpath)
 	}
```

Comment:
I don't see any instance in which it could be greater than 1 based of the way selectors work.

It should equal exactly 1, assuming the entire path was resolved -- if it wasn't, then it will have 0 nodes, indicating a failure.